# Semiconductor Equipment Reliability and Anomaly Detection

## Step 1: Data Ingestion and Cleaning

For this project, I am using **SECOM Manufacturing Dataset (UCI)**. This dataset consists of 590 sensors data. Since a complex modern semi-conoductor manufacturing process is normally under consistent surveillance throught eh monitoring of signal/variables collectes from sensors and process measurement points. 

Some of these signals are trivial. It could consist of irrelevant information or noise. 

My goal for this step is to load law sensor data, understand its structure, and handle any missing values.

In [2]:
import pandas as pd
import numpy as np

# Loading the dataset
df = pd.read_csv('C:/Users/Namita/Desktop/Projects/Semiconductor/data/uci-secom.csv')

print(f"Shape: {df.shape}")
df.head()

Shape: (1567, 592)


,Time,0,1,2,3,4,5,6,7,8,...,581,582,583,584,585,586,587,588,589,Pass/Fail
0,2008-07-19 11:55:00,3030.93,2564.00,2187.7333,1411.1265,1.3602,100.0,97.6133,0.1242,1.5005,...,NaN,0.5005,0.0118,0.0035,2.3630,NaN,NaN,NaN,NaN,-1
1,2008-07-19 12:32:00,3095.78,2465.14,2230.4222,1463.6606,0.8294,100.0,102.3433,0.1247,1.4966,...,208.2045,0.5019,0.0223,0.0055,4.4447,0.0096,0.0201,0.0060,208.2045,-1
2,2008-07-19 13:17:00,2932.61,2559.94,2186.4111,1698.0172,1.5102,100.0,95.4878,0.1241,1.4436,...,82.8602,0.4958,0.0157,0.0039,3.1745,0.0584,0.0484,0.0148,82.8602,1
3,2008-07-19 14:43:00,2988.72,2479.90,2199.0333,909.7926,1.3204,100.0,104.2367,0.1217,1.4882,...,73.8432,0.4990,0.0103,0.0025,2.0544,0.0202,0.0149,0.0044,73.8432,-1
4,2008-07-19 15:22:00,3032.24,2502.87,2233.3667,1326.5200,1.5334,100.0,100.3967,0.1235,1.5031,...,NaN,0.4800,0.4766,0.1045,99.3032,0.0202,0.0149,0.0044,73.8432,-1


In [3]:
# Checking the Pass/Fail values
print(df['Pass/Fail'].value_counts())

Pass/Fail
-1    1463
 1     104
Name: count, dtype: int64


With the result above, we see that the Pass values are over ten times the Fail values. This is called class imbalance. This means the the model could predict 'pass' everytime and be over 90% accurate while missing every fail result. This is a problem.

In [8]:
# Checking missing values
print(df.isnull().sum())
print(df.isnull().mean()*100)
print(f'\nTotal missing values: {df.isnull().sum().sum()}')
print(f'Missing values percentage: {df.isnull().mean().mean()*100}')

Time          0
0             6
1             7
2            14
3            14
             ..
586           1
587           1
588           1
589           1
Pass/Fail     0
Length: 592, dtype: int64
Time         0.000000
0            0.382897
1            0.446713
2            0.893427
3            0.893427
               ...   
586          0.063816
587          0.063816
588          0.063816
589          0.063816
Pass/Fail    0.000000
Length: 592, dtype: float64

Total missing values: 41951
Missing values percentage: 4.522219251798065


Although, there are a lot of missing values, the overall percentage of missing values is just 4.5%